domain/src/test/java/org/sagebionetworks/research/domain/TreeNavigatorTests.java

```diff
@@ -0,0 +1,334 @@
+    }
+
+    @Test
+    public void testProgress_MarkersAndSections_Completion() {
+        List<Result> stepHistory = new ArrayList<>();
+        for (int i = 0; i < 3; i++) {
+            stepHistory.add(mockResult(TEST_STEPS.get(i)));
+        }
+
+        SectionStep step4 = (SectionStep) TEST_STEPS.get(4);
+        stepHistory.add(mockTaskResult(step4.getIdentifier(), step4.getSteps()));
+        SectionStep step5 = (SectionStep) TEST_STEPS.get(5);
+        stepHistory.add(mockTaskResult(step5.getIdentifier(), step5.getSteps()));
+        SectionStep step6 = (SectionStep) TEST_STEPS.get(6);
+        stepHistory.add(mockTaskResult(step6.getIdentifier(), step6.getSteps()));
+        stepHistory.add(mockResult(TEST_STEPS.get(7)));
+        TaskResult taskResult = mockTaskResultFromResults("task", stepHistory);
+        Task.Progress progress = TEST_NAVIGATOR.getProgress(TEST_STEPS.get(8), taskResult);
+        assertNull(progress);
+    }
+
+    // MARK: test getPreviousStep()
+
+    @Test
+    public void testBack_From5X() {
+        List<Result> stepHistory = new ArrayList<>();
+        for (int i = 0; i < 3; i++) {
+            stepHistory.add(mockResult(TEST_STEPS.get(i)));
+        }
+
+        SectionStep step4 = (SectionStep) TEST_STEPS.get(4);
+        stepHistory.add(mockTaskResult(step4.getIdentifier(), step4.getSteps()));
+        SectionStep step5 = (SectionStep) TEST_STEPS.get(5);
+        stepHistory.add(mockTaskResult(step5.getIdentifier(), new ArrayList<Step>()));
+        TaskResult taskResult = mockTaskResultFromResults("task", stepHistory);
+        Step previousStep = TEST_NAVIGATOR.getPreviousStep(step5.getSteps().get(0), taskResult);
+        assertNotNull(previousStep);
+        assertEquals("step4.C", previousStep.getIdentifier());
+    }
+
+    @Test
+    public void testBack_From5Z() {
+        List<Result> stepHistory = new ArrayList<>();
+        for (int i = 0; i < 3; i++) {
+            stepHistory.add(mockResult(TEST_STEPS.get(i)));
+        }
+
+        SectionStep step4 = (SectionStep) TEST_STEPS.get(4);
+        stepHistory.add(mockTaskResult(step4.getIdentifier(), step4.getSteps()));
+        SectionStep step5 = (SectionStep) TEST_STEPS.get(5);
+        stepHistory.add(mockTaskResult(step5.getIdentifier(), step5.getSteps().subList(0, 2)));
+        TaskResult taskResult = mockTaskResultFromResults("task", stepHistory);
+        Step previousStep = TEST_NAVIGATOR.getPreviousStep(step5.getSteps().get(2), taskResult);
+        assertNotNull(previousStep);
+        assertEquals("step5.Y", previousStep.getIdentifier());
+    }
+
+    @Test
+    public void testBack_From3() {
+        TaskResult taskResult = mockTaskResult("task", TEST_STEPS.subList(1, 3));
+        Step previousStep = TEST_NAVIGATOR.getPreviousStep(TEST_STEPS.get(3), taskResult);
+        assertNotNull(previousStep);
+        assertEquals("step2", previousStep.getIdentifier());
+    }
+
+    // MARK: test getNextStep()
```

Comment:
These help with code folding and also in the code navigation pane
https://www.jetbrains.com/help/idea/code-folding.html#using_folding_comments

I prefer the Visual Studio one myself

Reply:
Awesome, I am going to add these to the files and push another commit.